### ISSUE TITLE
gRPC compatibility?

### ISSUE THREAD

**Issue by ineiti** [dedis/cothority#1824]
ByzGen asked whether it would be possible to have a more standard client-interface:

- gRPC - would be quite close to what we already have
- REST/JSON - would open up a bigger set of possible interactions

Any thoughts on feasibility of these? Either on the standard port, or on yet another port?

**Comment by kc1212** [comment]
Initial investigation results:

gRPC is similar to protobuf except that it additionally generates functions for making the RPC call. What you write are messages and service definitions like this:

```protobuf
service Greeter {
  rpc SayHello (HelloRequest) returns (HelloReply) {}
}

message HelloRequest {
  string name = 1;
}

message HelloReply {
  string message = 1;
}
```

Then it'll generate functions which you can use to create clients and servers, and send messages. For example:
```go
// CLIENT
c := pb.NewGreeterClient(conn)
// setup context 
c.SayHello(ctx, &pb.HelloRequest{Name: "hello world"})

// SERVER
type server struct{}
func (s *server) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
	log.Printf("Received: %v", in.Name)
	return &pb.HelloReply{Message: "Hello " + in.Name}, nil
}
// register server struct to the gRPC server
s := grpc.NewServer()
pb.RegisterGreeterServer(s, &server{})
```

Just like protobuf, the right way to use gRPC is write these messages and definitions in `.proto` files and then generate client and server code in various programming languages. However, this is not the model that cothority is using so we need to think of another way of using gRPC.

If we stick to our current approach, we need to generate gRPC-style `.proto` files, like `service Greeter` in the example above. Which means we need to parse our service handlers. We also need a translation layer and gRPC server in onet that converts requests to gRPC to the appropriate handlers.

My current feeling is that we shouldn't invest in more non-standard tech, that is, generating gRPC protobuf files from go source code.

**Comment by ineiti** [comment]
> My current feeling is that we shouldn't invest in more non-standard tech, that is, generating gRPC protobuf files from go source code.

Yes, definitely. I already though what would be needed to do the 'correct' .proto -> go conversion. In fact, we could probably make it quite painless by simply _embedding_ the created protobuf-structures in our own. The only downside would be that it would not correctly parse dependant structures: so if you have a `onet.Roster` field, I don't know how you could turn it into a `onet.Roster` instead of a `protobuf.Roster`. In Javascript, Gaylor wrote nice wrapper-methods that allow to do it. But I don't know if this is possible in go.

**Comment by kc1212** [comment]
first experiment: https://github.com/dedis/onet/commit/cf54dc9022c9b09302e9777b8d24398fd9fc7e67

**Comment by kc1212** [comment]
first experiment: https://github.com/dedis/onet/commit/cf54dc9022c9b09302e9777b8d24398fd9fc7e67

Running something like `curl --data '{"Version": 1, "Key": "ZW5jb2Rpbmc=", "ID": "ZW5jb2Rpbmc="}' http://127.0.0.1:7771/v3/ByzCoin/GetProof` will execute the right handler (if the Key and ID are valid and `RegisterRESTHandler` is called on `GetProof`).

**Comment by jeffallen** [comment]
Answers (my opinions):
1. no, because
2. no , if we do REST, we do it right. Idempotent calls are GET (this solves the `--data '{}'` thing as well.
3. I agree: new handler
4. REST interfaces use HTTP error codes as much as possible. See Github API for a nice example: https://developer.github.com/v3/#client-errors

https://api.github.com/users/jeffallen/keys -> code 200, body is JSON result
https://api.github.com/users/jeffallen/keys/extra -> code 404, body is JSON error:
```
{
  "message": "Not Found",
  "documentation_url": "https://developer.github.com/v3"
}
```

**Comment by jeffallen** [comment]
The infrastructure for this is dedis/onet#548, available in onet v3.0.13. Edited the title to indicate this issue tracks making use of the new system for OCS.

**Comment by nkcr** [comment]
@ineiti Is this still an interesting feature for Byzgen? 
Should we implement it for v4?

**Comment by ineiti** [comment]
ByzGen is currently undergoing restructuring, I have no news for the moment.

I had a short discussion today with @Gilthoniel on how to change the conode<->conode communication. Perhaps there is a possibility to build something in common for those two use-cases?

**Comment by nkcr** [comment]
> Perhaps there is a possibility to build something in common for those two use-cases

Yes, but I guess that for the sake of performance a server<->server communication would be better with gRpc rather than REST.

**Comment by ineiti** [comment]
gRPC uses HTTP/2 for transportation - so it is a _possibility_ to put both on the same port. Then again, perhaps it's not a good idea ;)

Just a thought.

And for REST in the client->conode communication, you'd have to make sure that the _stream_ function still works, and that the connection is reused.
